# Assigment 7
This assigment will be graded if everything works well. I will run the script as once and everything should be done without errors and mistakes. I should be able to run your scripts in my computer and get all the results. USE RELATIVE PATHS. An error or exception or anything that breaks the code will means NO GRADE (0). Additionally, you are not able to modify any file handly. It also means NO GRADE (0). Comment everything you think will help others read your script. We expect 0 errors using GitHub. Everything will be graded!

ASK EVERYTHING! WE ARE HERE TO HELP YOU!

GET YOUR GOOGLE API AND TOKEN. YOU WILL NEED THEM TO DO THIS TASK.

- Import Data from this url. This dataset is in PDF format. You have to convert to PandasDataFrame and keep only the BBVA offices in LIMA.
- Use GoogleMaps API and geocode all the BBVA offices. For those offices that Google API gets no information, use internet and get the latitude and longitude handly and add them to dataset.
- Use Google API to find the driving time (best guess) from all the group members' address and all the LIMA BBVA offices.
- Finally, you have to give a report which offices are the most closest and furthest to every group member's address.

In [1]:
# Packages
import pandas as pd
import os
import urllib.request, json, csv
import numpy as np
from tqdm import tqdm_notebook as tqdm
# For sending GET requests from the API
import requests
# For saving access tokens and for file management when creating and adding to the dataset
import os
# For dealing with json responses we receive from the API
import json
# For displaying the data after
import pandas as pd
# For saving the response data in CSV format
import csv
# For parsing the dates received from twitter in readable formats
import datetime
import dateutil.parser
import unicodedata
#To add wait time between requests
import time
import requests
#maps
import googlemaps

## 1. First we import the pdf file into pandas and only keep desired rows

Import Data from this url. This dataset is in PDF format. You have to convert to PandasDataFrame and keep only the BBVA offices in LIMA.

Yo lo convertí a excel por fuera y lo subí directo de mis documentos.
Faltaría:
- Encontrar cómo subirlo de PDF para que identifique las tablas
- Usar relative paths

In [2]:
file_path = "/Users/alejandroboyco/Documents/Python/Oficinas-BBVA-abiertas-23.10.20.xlsx"
df = pd.read_excel(file_path)

In [3]:
#Convertir la primera fila en el nombre de las columnas

df.rename(columns=df.iloc[0], inplace = True)
df.drop(df.index[0], inplace = True)
df

OFICINA  \
1               BAGUA GRANDE   
2                   CHIMBOTE   
3                     HUARAZ   
4         MEGAPLAZA CHIMBOTE   
5    MERCADO MODELO CHIMBOTE   
..                       ...   
302                    TACNA   
303                   TUMBES   
304               7 DE JUNIO   
305      OPEN PLAZA PUCALLPA   
306                 PUCALLPA   

                                             DIRECCION      DISTRITO  \
1                                 AV. CHACHAPOYAS 2044  BAGUA GRANDE   
2               AV. FRANCISCO BOLOGNESI 696 - CHIMBOTE      CHIMBOTE   
3         CALLE 2 ESQ. JR. JOSE SUCRE - PLAZA DE ARMAS        HUARAZ   
4    AV. VICTOR R. HAYA DE LA TORRE 4694, LC B4 Y B...      CHIMBOTE   
5    JR. ELIAS AGUIRRE 498, ESQ. CON JR. LADISLAO E...      CHIMBOTE   
..                                                 ...           ...   
302                         AV. SAN MARTIN 665 - TACNA         TACNA   
303                                    JR. BOLIVAR 237        TUMBES   
304                           JIRÓN 7 DE JUNIO N.- 201      CALLERIA   
305  AV. CENTENARIO 2086 Y 2192, JR. LAUREANO DEL Á...   YARINACOCHA   
306                     ESQUINA UCAYALI Y RAYMONDI 699      CALLERIA   

             PROVINCIA DEPARTAMENTO  
1            UTCUBAMBA     AMAZONAS  
2                SANTA       ANCASH  
3               HUARAZ       ANCASH  
4                SANTA       ANCASH  
5                SANTA       ANCASH  
..                 ...          ...  
302              TACNA        TACNA  
303             TUMBES       TUMBES  
304  CORONEL\nPORTILLO      UCAYALI  
305  CORONEL\nPORTILLO      UCAYALI  
306   CORONEL PORTILLO      UCAYALI  

[306 rows x 5 columns]

In [4]:
#nos quedamos solo con el departamento de Lima

df = df.loc[df['DEPARTAMENTO'] == "LIMA"]
df = df.reset_index(drop = True)
df

OFICINA  \
0               CAMINO REAL   
1               28 DE JULIO   
2          ABANCAY-CONGRESO   
3                AEROPUERTO   
4              ANGAMOS ESTE   
..                      ...   
201     NESTLE-DONOFRIO 164   
202  BIENESTAR DE LA MARINA   
203                 SEDAPAL   
204         ESPECIAL SENATI   
205          VILLA INFANTAS   

                                             DIRECCION       DISTRITO  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    PROVINCIA DEPARTAMENTO  
0        LIMA         LIMA  
1        LIMA         LIMA  
2        LIMA         LIMA  
3        LIMA         LIMA  
4        LIMA         LIMA  
..        ...          ...  
201      LIMA         LIMA  
202      LIMA         LIMA  
203      LIMA         LIMA  
204      LIMA         LIMA  
205      LIMA         LIMA  

[206 rows x 5 columns]

## 2. Then we find the coordinates for each BBVA office

Use GoogleMaps API and geocode all the BBVA offices. For those offices that Google API gets no information, use internet and get the latitude and longitude handly and add them to dataset.


In [5]:
df['direc'] = df['DIRECCION'] + " " + df['DISTRITO'] + " " + "BBVA"
df
#By adding strings from "DISTRITO", there is more information in the adress and we reduce the NANs from 39 to 7.
#Then, by adding "BBVA" to the string we reduce the missing values from 6 to 3.

OFICINA  \
0               CAMINO REAL   
1               28 DE JULIO   
2          ABANCAY-CONGRESO   
3                AEROPUERTO   
4              ANGAMOS ESTE   
..                      ...   
201     NESTLE-DONOFRIO 164   
202  BIENESTAR DE LA MARINA   
203                 SEDAPAL   
204         ESPECIAL SENATI   
205          VILLA INFANTAS   

                                             DIRECCION       DISTRITO  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    PROVINCIA DEPARTAMENTO                                              direc  
0        LIMA         LIMA                AV. CAMINO REAL 355 SAN ISIDRO BBVA  
1        LIMA         LIMA                      AV. 28 DE JULIO 775 LIMA BBVA  
2        LIMA         LIMA                      AV. ABANCAY 260-262 LIMA BBVA  
3        LIMA         LIMA  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A C...  
4        LIMA         LIMA  Av. Angamos Este 215 - 227, esquina con Av. Pe...  
..        ...          ...                                                ...  
201      LIMA         LIMA                       AV. VENEZUELA 2580 LIMA BBVA  
202      LIMA         LIMA          AV. VENEZUELA S/N CDRA 34 SAN MIGUEL BBVA  
203      LIMA         LIMA  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...  
204      LIMA         LIMA       AV. ALFREDO MENDIOLA 3520 INDEPENDENCIA BBVA  
205      LIMA         LIMA  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...  

[206 rows x 6 columns]

In [6]:
direc = df["direc"]
direc.head()

0                  AV. CAMINO REAL 355 SAN ISIDRO BBVA
1                        AV. 28 DE JULIO 775 LIMA BBVA
2                        AV. ABANCAY 260-262 LIMA BBVA
3    CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A C...
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...
Name: direc, dtype: object

In [7]:
coord = np.zeros(shape=( len(direc), 2), dtype =float)
i=0

gmaps = googlemaps.Client(key='AIzaSyD_4E6Hd-fYECy3mZ4asxN23JjIstvLdoE')

# Loop over lists
for d in tqdm(direc):   

    # Geocoding an address
    geocode_result = gmaps.geocode( d , region = 'pe')
    
    # Check the len of result
    if len(geocode_result)==0 :
        coord[i][0] = np.nan
        coord[i][1] = np.nan

    
    # Get info
    else :
        coord[i][0] = geocode_result[0]['geometry']['location']['lat']
        coord[i][1] = geocode_result[0]['geometry']['location']['lng']
        
    i=i+1

/var/folders/79/q_j4g_550cd5m0w3kcvjwhjw0000gn/T/ipykernel_11285/1208104145.py:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for d in tqdm(direc):


  0%|          | 0/206 [00:00<?, ?it/s]

In [8]:
#Create the dataframe with the new coordenates
coords = pd.DataFrame( coord.tolist() , columns = ["Lat", "Lon"] )
coords['direc'] = direc
coords

Lat        Lon                                              direc
0   -12.097362 -77.036099                AV. CAMINO REAL 355 SAN ISIDRO BBVA
1   -12.064030 -77.037961                      AV. 28 DE JULIO 775 LIMA BBVA
2   -12.047955 -77.026997                      AV. ABANCAY 260-262 LIMA BBVA
3   -12.023538 -77.104397  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A C...
4   -12.113476 -77.028582  Av. Angamos Este 215 - 227, esquina con Av. Pe...
..         ...        ...                                                ...
201 -12.062448 -77.093581                       AV. VENEZUELA 2580 LIMA BBVA
202 -12.077434 -77.093414          AV. VENEZUELA S/N CDRA 34 SAN MIGUEL BBVA
203 -12.033125 -76.983811  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...
204 -11.994069 -77.063041       AV. ALFREDO MENDIOLA 3520 INDEPENDENCIA BBVA
205 -11.959300 -77.076025  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...

[206 rows x 3 columns]

In [9]:
#Now we merge the coordenates dataframe with the original dataframe into the final df
df_final = df.merge(coords, left_on='direc', right_on='direc', how='left')
df_final.drop('direc', inplace=True, axis=1)
df_final.index.name = 'id'
df_final.reset_index(inplace=True)
df_final

id                 OFICINA  \
0      0             CAMINO REAL   
1      1             28 DE JULIO   
2      2        ABANCAY-CONGRESO   
3      3              AEROPUERTO   
4      4            ANGAMOS ESTE   
..   ...                     ...   
201  201     NESTLE-DONOFRIO 164   
202  202  BIENESTAR DE LA MARINA   
203  203                 SEDAPAL   
204  204         ESPECIAL SENATI   
205  205          VILLA INFANTAS   

                                             DIRECCION       DISTRITO  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    PROVINCIA DEPARTAMENTO        Lat        Lon  
0        LIMA         LIMA -12.097362 -77.036099  
1        LIMA         LIMA -12.064030 -77.037961  
2        LIMA         LIMA -12.047955 -77.026997  
3        LIMA         LIMA -12.023538 -77.104397  
4        LIMA         LIMA -12.113476 -77.028582  
..        ...          ...        ...        ...  
201      LIMA         LIMA -12.062448 -77.093581  
202      LIMA         LIMA -12.077434 -77.093414  
203      LIMA         LIMA -12.033125 -76.983811  
204      LIMA         LIMA -11.994069 -77.063041  
205      LIMA         LIMA -11.959300 -77.076025  

[206 rows x 8 columns]

### Now we replace NAs by hand

In [10]:
isnan = df_final[df_final.isna().any(axis=1)]
len(isnan)
#There are 3 cases with missing values: Mala, Naciones Unidas & Strip Center Barranca

3

In [11]:
isnan

id                OFICINA                  DIRECCION    DISTRITO  \
113  113                   MALA               JR. REAL 305        MALA   
127  127        NACIONES UNIDAS  Av. Jorge Chavez Cuadra 2  MIRAFLORES   
196  196  STRIP CENTER BARRANCA  CA. CASTILLA 370, LOCAL 1    BARRANCA   

    PROVINCIA DEPARTAMENTO  Lat  Lon  
113    CAÑETE         LIMA  NaN  NaN  
127      LIMA         LIMA  NaN  NaN  
196  BARRANCA         LIMA  NaN  NaN

In [12]:
#Oficina Mala
df_final.loc[113,'Lat']= -12.655853899007829
df_final.loc[113,'Lon']= -76.63487728412044
#Oficina Naciones Unidas
df_final.loc[127,'Lat']= -12.100351035450258
df_final.loc[127,'Lon']= -77.06362336201971
#Oficina Strip Center Barranca
df_final.loc[40,'Lat']= -10.752370606647778
df_final.loc[40,'Lon']= -77.75986317149847

## 3. Find driving time

Use Google API to find the driving time (best guess) from all the group members' address and all the LIMA BBVA offices.

Les puse direcciones al azar. Sería cuestión que lo editen.

#### Group members' address
- Alejandro: -12.114876114474438, -77.03866001449956
- EDITAR Steph: -12.069439763000675, -77.07819124570732
- EDITAR Gabriela: -12.104845219259536, -76.95576067331922
- EDITAR Juan Carlos: -12.023947924213122, -77.04884398480819

In [13]:
#We subset only coordenates and transform them to strings instead of floats
bbva = df_final.iloc[:, [6,7]].applymap(str)

In [38]:
#Create a list of "destinations"
bbva['destino'] = bbva['Lat'] + "," + bbva['Lon']
destino = bbva['destino'].tolist()
destino

['-12.0973616,-77.0360994',
 '-12.0640305,-77.0379609',
 '-12.0479547,-77.02699679999999',
 '-12.0235384,-77.10439749999999',
 '-12.1134759,-77.028582',
 '-12.1120524,-77.0117853',
 '-12.0719308,-77.011235',
 '-12.1036288,-77.0307414',
 '-12.0716113,-77.0374212',
 '-12.0915794,-77.03373189999999',
 '-12.1140109,-77.0480248',
 '-12.0307591,-77.0848436',
 '-12.0807507,-77.0665087',
 '-12.0965008,-77.0020389',
 '-12.0470329,-76.93830129999999',
 '-10.752491,-77.7598701',
 '-12.1473578,-77.02133959999999',
 '-12.0370867,-77.1372552',
 '-12.0534252,-77.0958897',
 '-12.1266464,-77.01623',
 '-12.0722545,-77.0661447',
 '-12.0358864,-77.0933675',
 '-12.1217533,-77.0117212',
 '-12.0622538,-77.0665705',
 '-12.0088873,-77.0602449',
 '-12.1285546,-76.9993054',
 '-12.104926,-76.9707944',
 '-12.12913,-76.9841241',
 '-12.0852851,-77.03221669999999',
 '-12.0954325,-77.02623779999999',
 '-12.076778,-77.083096',
 '-12.0754679,-77.064285',
 '-12.0608102,-77.143934',
 '-12.0447042,-76.9343444',
 '-12.04811

In [39]:
#Alejandro
df_driving = pd.DataFrame (destino, columns = ['destination'])
df_driving['origin'] = "-12.114876114474438,-77.03866001449956"
df_driving

destination                                  origin
0           -12.0973616,-77.0360994  -12.114876114474438,-77.03866001449956
1           -12.0640305,-77.0379609  -12.114876114474438,-77.03866001449956
2    -12.0479547,-77.02699679999999  -12.114876114474438,-77.03866001449956
3    -12.0235384,-77.10439749999999  -12.114876114474438,-77.03866001449956
4            -12.1134759,-77.028582  -12.114876114474438,-77.03866001449956
..                              ...                                     ...
201         -12.0624481,-77.0935808  -12.114876114474438,-77.03866001449956
202         -12.0774344,-77.0934137  -12.114876114474438,-77.03866001449956
203         -12.0331248,-76.9838113  -12.114876114474438,-77.03866001449956
204         -11.9940687,-77.0630414  -12.114876114474438,-77.03866001449956
205  -11.9593005,-77.07602539999999  -12.114876114474438,-77.03866001449956

[206 rows x 2 columns]

In [40]:
#Steph
df_driving2 = pd.DataFrame (destino, columns = ['destination'])
df_driving2['origin'] = "-12.069439763000675,-77.07819124570732"
df_driving2

destination                                  origin
0           -12.0973616,-77.0360994  -12.069439763000675,-77.07819124570732
1           -12.0640305,-77.0379609  -12.069439763000675,-77.07819124570732
2    -12.0479547,-77.02699679999999  -12.069439763000675,-77.07819124570732
3    -12.0235384,-77.10439749999999  -12.069439763000675,-77.07819124570732
4            -12.1134759,-77.028582  -12.069439763000675,-77.07819124570732
..                              ...                                     ...
201         -12.0624481,-77.0935808  -12.069439763000675,-77.07819124570732
202         -12.0774344,-77.0934137  -12.069439763000675,-77.07819124570732
203         -12.0331248,-76.9838113  -12.069439763000675,-77.07819124570732
204         -11.9940687,-77.0630414  -12.069439763000675,-77.07819124570732
205  -11.9593005,-77.07602539999999  -12.069439763000675,-77.07819124570732

[206 rows x 2 columns]

In [41]:
#Gabriela
df_driving3 = pd.DataFrame (destino, columns = ['destination'])
df_driving3['origin'] = "-12.104845219259536,-76.95576067331922"
df_driving3

destination                                  origin
0           -12.0973616,-77.0360994  -12.104845219259536,-76.95576067331922
1           -12.0640305,-77.0379609  -12.104845219259536,-76.95576067331922
2    -12.0479547,-77.02699679999999  -12.104845219259536,-76.95576067331922
3    -12.0235384,-77.10439749999999  -12.104845219259536,-76.95576067331922
4            -12.1134759,-77.028582  -12.104845219259536,-76.95576067331922
..                              ...                                     ...
201         -12.0624481,-77.0935808  -12.104845219259536,-76.95576067331922
202         -12.0774344,-77.0934137  -12.104845219259536,-76.95576067331922
203         -12.0331248,-76.9838113  -12.104845219259536,-76.95576067331922
204         -11.9940687,-77.0630414  -12.104845219259536,-76.95576067331922
205  -11.9593005,-77.07602539999999  -12.104845219259536,-76.95576067331922

[206 rows x 2 columns]

In [42]:
#Juan Carlos
df_driving4 = pd.DataFrame (destino, columns = ['destination'])
df_driving4['origin'] = "-12.023947924213122,-77.04884398480819"
df_driving4

destination                                  origin
0           -12.0973616,-77.0360994  -12.023947924213122,-77.04884398480819
1           -12.0640305,-77.0379609  -12.023947924213122,-77.04884398480819
2    -12.0479547,-77.02699679999999  -12.023947924213122,-77.04884398480819
3    -12.0235384,-77.10439749999999  -12.023947924213122,-77.04884398480819
4            -12.1134759,-77.028582  -12.023947924213122,-77.04884398480819
..                              ...                                     ...
201         -12.0624481,-77.0935808  -12.023947924213122,-77.04884398480819
202         -12.0774344,-77.0934137  -12.023947924213122,-77.04884398480819
203         -12.0331248,-76.9838113  -12.023947924213122,-77.04884398480819
204         -11.9940687,-77.0630414  -12.023947924213122,-77.04884398480819
205  -11.9593005,-77.07602539999999  -12.023947924213122,-77.04884398480819

[206 rows x 2 columns]

In [43]:
df_driving_final = pd.concat([df_driving, df_driving2, df_driving3, df_driving4]).reset_index(drop=True)
df_driving_final

destination                                  origin
0           -12.0973616,-77.0360994  -12.114876114474438,-77.03866001449956
1           -12.0640305,-77.0379609  -12.114876114474438,-77.03866001449956
2    -12.0479547,-77.02699679999999  -12.114876114474438,-77.03866001449956
3    -12.0235384,-77.10439749999999  -12.114876114474438,-77.03866001449956
4            -12.1134759,-77.028582  -12.114876114474438,-77.03866001449956
..                              ...                                     ...
819         -12.0624481,-77.0935808  -12.023947924213122,-77.04884398480819
820         -12.0774344,-77.0934137  -12.023947924213122,-77.04884398480819
821         -12.0331248,-76.9838113  -12.023947924213122,-77.04884398480819
822         -11.9940687,-77.0630414  -12.023947924213122,-77.04884398480819
823  -11.9593005,-77.07602539999999  -12.023947924213122,-77.04884398480819

[824 rows x 2 columns]

In [44]:
df_driving_final['combination'] = list(range(1,len(df_driving_final)+1))
df_driving_final

destination                                  origin  \
0           -12.0973616,-77.0360994  -12.114876114474438,-77.03866001449956   
1           -12.0640305,-77.0379609  -12.114876114474438,-77.03866001449956   
2    -12.0479547,-77.02699679999999  -12.114876114474438,-77.03866001449956   
3    -12.0235384,-77.10439749999999  -12.114876114474438,-77.03866001449956   
4            -12.1134759,-77.028582  -12.114876114474438,-77.03866001449956   
..                              ...                                     ...   
819         -12.0624481,-77.0935808  -12.023947924213122,-77.04884398480819   
820         -12.0774344,-77.0934137  -12.023947924213122,-77.04884398480819   
821         -12.0331248,-76.9838113  -12.023947924213122,-77.04884398480819   
822         -11.9940687,-77.0630414  -12.023947924213122,-77.04884398480819   
823  -11.9593005,-77.07602539999999  -12.023947924213122,-77.04884398480819   

     combination  
0              1  
1              2  
2              3  
3              4  
4              5  
..           ...  
819          820  
820          821  
821          822  
822          823  
823          824  

[824 rows x 3 columns]

In [45]:
#Create the three lists: combination, origin and destination
combination = df_driving_final['combination'].tolist()
origin = df_driving_final['origin'].tolist()
destination = df_driving_final['destination'].tolist()

In [47]:
data_distance = {} 

In [63]:
# Loop to generate info about geolocations
duration_info = np.zeros(shape=(len(combination),3), dtype =float)
i=0

for c,o,d in tqdm(list(zip(combination, origin, destination))):
    try:
        # Google MapsDdirections API endpoint
        endpoint = 'https://maps.googleapis.com/maps/api/directions/json?'

    ## Fixed Parameters
        # Paramaters
        traffic_model = 'best_guess'  

        # Departure time
        departure_time = 'now'

        # driving, walking, biclycling, transit
        mode = 'driving'

        # key
        api_key = 'AIzaSyD_4E6Hd-fYECy3mZ4asxN23JjIstvLdoE'

        # region to look for (spain= es , germany = de , Switzerland= swiss)
        region = 'pe'

     ## Parameters
        # Origin
        origin = o
        
        # Destinations
        destination = d
        
        #Building the URL for the request
        nav_request = 'origin={}&destination={}&departure_time={}&traffic_model={}&mode={}&region={}&key={}'.format(origin , 
                        destination , departure_time , traffic_model , mode, region, api_key)        
        
        # Concatenate strings
        request = endpoint + nav_request

        #Sends the request and reads the response.
        response = urllib.request.urlopen(request).read()

        #Loads response as JSON
        directions = json.loads(response)

        legs = directions['routes'][0]['legs'][0]
        
        duration_info[i][0] = c
        duration_info[i][1] = float(re.sub("[^0-9.]", "", legs['duration']['text']))
        duration_info[i][2] = legs['duration']['value']
        
        duration_info[i][0] = c
        duration_info[i][1] = tiempo
        
        i=i+1

        my_keys = ['duration']
        info = { my_key: legs[my_key] for my_key in my_keys }
        data_distance[c] = info 
    
    except Exception as e:
        
        duration_info[i][0] = c
        duration_info[i][1] = "nan"
        duration_info[i][2] = "nan"
        
        i=i+1
        
        data_distance[c] = {'duration': {'text': 'nan', 'value': 0}}
        
        print(c)

/var/folders/79/q_j4g_550cd5m0w3kcvjwhjw0000gn/T/ipykernel_11285/3327988368.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for c,o,d in tqdm(list(zip(combination, origin, destination))):


  0%|          | 0/1 [00:00<?, ?it/s]

1


In [64]:
duration_info

array([[ 1., nan, nan],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       ...,
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.]])

In [62]:
distance_api = pd.DataFrame(duration_info, columns = ['Combination', 'Duration_min', 'Duration_seconds'])
distance_api

Combination  Duration_min  Duration_seconds
0            1.0           NaN               NaN
1            0.0           0.0               0.0
2            0.0           0.0               0.0
3            0.0           0.0               0.0
4            0.0           0.0               0.0
..           ...           ...               ...
819          0.0           0.0               0.0
820          0.0           0.0               0.0
821          0.0           0.0               0.0
822          0.0           0.0               0.0
823          0.0           0.0               0.0

[824 rows x 3 columns]

Por alguna razón no sale, pero probando con uno solo (sin loop) sí:

In [55]:
duration_info = np.zeros(shape=(len(combination),2), dtype =float)

endpoint = 'https://maps.googleapis.com/maps/api/directions/json?'

## Fixed Parameters
# Paramaters
traffic_model = 'best_guess'  

# Departure time
departure_time = 'now'

# driving, walking, biclycling, transit
mode = 'driving'

# key
api_key = 'AIzaSyD_4E6Hd-fYECy3mZ4asxN23JjIstvLdoE'

# region to look for (spain= es , germany = de , Switzerland= swiss)
region = 'pe'

## Parameters
# Origin
origin = '-12.114876114474438,-77.03866001449956'
        
# Destinations
destination = '-12.0235384,-77.10439749999999'

nav_request = 'origin={}&destination={}&departure_time={}&traffic_model={}&mode={}&region={}&key={}'.format(origin ,
                            destination , departure_time , traffic_model , mode, region, api_key)        
        
# Concatenate strings
request = endpoint + nav_request

#Sends the request and reads the response.
response = urllib.request.urlopen(request).read()

#Loads response as JSON
directions = json.loads(response)

legs = directions['routes'][0]['legs'][0]
tiempo = legs['duration']['value']

In [56]:
tiempo

2131

## 4. Report offices closets and furthest from each group members' adress

Se me ocurre que podríamos crear un dataframe en base a lo anterior (cuando se arregle) que sea una columna por miembro del grupo con el tiempo a cada una de las 205 oficinas, y hacemos un merge con la df_final. Luego que nos dé la fila donde está el valor mínimo y máximo para cada una de las 4 columnas y así identificamos las oficinas.